skr/webhook/internal/handler.go

```diff
@@ -0,0 +1,343 @@
+package internal
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"reflect"
+
+	"github.com/go-logr/logr"
+	"github.com/kyma-project/kyma-watcher/kcp/pkg/types"
+	"github.com/kyma-project/kyma-watcher/skr/pkg/config"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Handler struct {
+	Client client.Client
+	Logger logr.Logger
+}
+
+type validateResource struct {
+	errorOccurred bool
+	allowed       bool
+	message       string
+	status        string
+}
+
+type responseInterface interface {
+	isEmpty() bool
+}
+
+type Resource struct {
+	metav1.GroupVersionKind
+	Status bool `json:"status"`
+	Spec   bool `json:"spec"`
+}
+
+type Metadata struct {
+	Name      string            `json:"name"`
+	Namespace string            `json:"namespace"`
+	Labels    map[string]string `json:"labels"`
+}
+
+func (m Metadata) isEmpty() bool {
+	return m.Name == ""
+}
+
+type ObjectWatched struct {
+	Metadata `json:"metadata"`
+	Spec     map[string]interface{} `json:"spec"`
+	Kind     string                 `json:"kind"`
+	Status   map[string]interface{} `json:"status"`
+}
+
+const (
+	AdmissionError       = "admission error:"
+	InvalidResource      = "invalid resource"
+	InvalidationMessage  = "invalidated from webhook"
+	ValidationMessage    = "validated from webhook"
+	writeFilePermissions = 0o600
+	defaultBufferSize    = 2048
+)
+
+//nolint:gochecknoglobals
+var UniversalDeserializer = serializer.NewCodecFactory(runtime.NewScheme()).UniversalDeserializer()
+
+func (h *Handler) Handle(writer http.ResponseWriter, req *http.Request) {
+	ctx := context.TODO()
+	kyma, resourceList := h.getKymaAndResourceListFromConfigMap(ctx)
+
+	// read incoming request to bytes
+	body, err := io.ReadAll(req.Body)
+	if err != nil {
+		h.httpError(writer, http.StatusInternalServerError, fmt.Errorf("%s %w",
+			AdmissionError, err))
+		return
+	}
+
+	// create admission review from bytes
+	admissionReview := h.getAdmissionRequestFromBytes(writer, body)
+	if admissionReview == nil {
+		return
+	}
+
+	resourceValidation := h.validateResources(writer, admissionReview, resourceList, kyma.GetName())
+
+	h.Logger.Info(fmt.Sprintf("incoming admission review for: %s", admissionReview.Request.Kind.Kind))
+
+	// store incoming request
+	//nolint:gosec
+	err = os.WriteFile("/tmp/request", body, writeFilePermissions)
```

Comment:
This can break on multiple requests at once, probably you should introduce an async writing with a channel or use a mutex here

Reply:
this was only for testing for the first iteration. I added a TODO to do so or implement a file mutex.